deployer/src/builder/oci/distribution/client.rs

```diff
@@ -0,0 +1,216 @@
+use http::Method;
+use oci_spec::image::*;
+
+use url::Url;
+
+use super::{super::error::*, super::image::digest::Digest, Name, Reference, StoredAuth};
+
+/// A client for `/v2/<name>/` API endpoint
+pub struct Client {
+    agent: reqwest::Client,
+    /// URL to registry server
+    url: Url,
+    /// Name of repository
+    name: Name,
+    /// Loaded authentication info from filesystem
+    auth: StoredAuth,
+    /// Cached token
+    token: Option<String>,
+}
+
+impl Client {
+    pub fn new(url: Url, name: Name) -> Result<Self> {
+        let auth = StoredAuth::load_all()?;
+        Ok(Client {
+            agent: reqwest::Client::new(),
+            url,
+            name,
+            auth,
+            token: None,
+        })
+    }
+
+    #[async_recursion::async_recursion]
```

Comment:
I've never seen this. What does it do? :star_struck:

Reply:
Hah, basically it's changing the return type of the function to a [`BoxedFuture`](https://docs.rs/futures/0.3.19/futures/future/type.BoxFuture.html).  A compiler error usually guides you to this when you write async recursive functions. An example can look like this:
```
 error[E0733]: recursion in an `async fn` requires boxing
--> src/main.rs:1:26
  |
1 | async fn fib(n : u32) -> u32 {
  |                          ^^^ recursive `async fn`
  |
  = note: a recursive `async fn` must be rewritten to return a boxed `dyn Future`.
  = note: consider using the `async_recursion` crate: https://crates.io/crates/async_recursion
```